### DISCUSSION TITLE
Custom header in image upload

### DISCUSSION THREAD

**Discussion by vesuvi88** [tneotia/html-editor-enhanced#166] [Q&A]
Hello everyone! I have the following problem:
In initialText I pass the html text <img alt="" src="https://myserv.com/storage/1.jpg ">. The server requires an authentication token in the header, since by default the token is not transmitted, the image cannot be uploaded. I thought I could use the Callback function onImageLinkInsert or onImageUploadError, but they don't respond to my text. How can I implement uploading an image with the right header?

**Comment by tneotia** [comment]
If you need to add the token for initialText, you will need to use the dart HTML package: https://pub.dev/packages/html

With this you can parse the HTML contents, extract the image nodes, change the URL to add the token to the URL, and then convert back to a string. Finally pass the converted string into initialText and it should work.

**Reply by vesuvi88** [reply]
Correction to my previous comment:
In initialText I pass the html text \<img alt="" src="https://myserv.com/storage/1.jpg">.

Thanks for the answer! But I need something else.
I need to add an authorization token to the http request header. For example: Image.network('https://myserv.com/storage/1.jpg ', headers: mycustomheader). Attached a screenshot of what should happen.
<img width="603" alt="Screenshot 2021-11-19 at 10 24 28" src="https://user-images.githubusercontent.com/19634035/142582343-7615f90b-6000-4960-a820-826178954ac7.png">

**Reply by tneotia** [reply]
Hmmm... since this library doesn't use Dart to load images under the hood (at the moment) this will be quite tricky. I see 2 potential options:

1. Use GET parameters with a token to replace the header for this functionality

`<img src="url/path/here?token=[TOKEN]">`

This token could be added into the src using the method I described in the initial post

2. Fetch the image beforehand with a Dart http request, and then using the method I described in the initial post, change the image src to base64 data, so something like this:

`<img src="data:[MIME TYPE];base64,[BASE64 ENCODED IMAGE]">`

**Reply by vesuvi88** [reply]
Thanks for the tips!

1. The server is waiting for the authorization token in the header field and does not perceive the parameter in the query string itself. I don't have access to the server code and I can't fix anything.

2. This option is also not suitable. As I understand it, you offer me the following: when transferring html text to html_editor_enhanced, convert the link to a base64 image and remember the correspondence of the link and base64. After editing, parse the html base64 images and put them back into link mode. Images can have a lot of weight and there can be a lot of them, so the size of the html file can be large and parsing can take a long time. Images are transmitted to and from the server by links for weight reasons.

**Reply by vesuvi88** [reply]
I think there is another solution. How can I add cookies to the flutter_inappwebview controller? I can pass the authorization token in cookies.

**Reply by tneotia** [reply]
Looks like you can use cookies without needing to interact with the webview itself: https://inappwebview.dev/docs/cookie-manager/basic-usage/

**Reply by AceChen1** [reply]
@vesuvi88  hi, I also meet the issue, may I know how can set jwt token to the cookies?

**Reply by vesuvi88** [reply]
Hi, @AceChen1 
I added an authorization token to cookies using a plugin https://pub.dev/packages/webview_cookie_manager.
```
final cookieManager = WebviewCookieManager();
await cookieManager.setCookies([
  Cookie('auth_key', token)
    ..domain = domainName
    ..expires = DateTime.now().add(const Duration(days: 10))
    ..httpOnly = false
]);
```
I tried adding cookies like here https://inappwebview.dev/docs/cookie-manager/basic-usage/, but it didn't work out for me.

**Reply by tneotia** [reply]
Nice find @vesuvi88! Good to see that this capability is possible.
